// igniteui-webcomponents-core/esm2015/lib/DatePickerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DatePickerDescription = /*@__PURE__*/ (() => {
    class DatePickerDescription extends Description {
        constructor() {
            super();
            this.m = new Date();
            this.l = new Date();
            this.y = null;
            this.ac = null;
            this.z = null;
            this.r = 0;
            this.aa = null;
            this.ab = null;
            this.ad = null;
            this.k = new Date();
            this.j = new Date();
            this.w = null;
            this.v = null;
            this.af = null;
            this.s = 0;
            this.ag = null;
            this.ah = null;
            this.ae = null;
            this.x = null;
            this.f = false;
            this.g = false;
            this.ai = null;
        }
        get_type() {
            return "DatePicker";
        }
        get type() {
            return this.get_type();
        }
        get value() {
            return this.m;
        }
        set value(a) {
            this.m = a;
            this.e("Value");
        }
        get today() {
            return this.l;
        }
        set today(a) {
            this.l = a;
            this.e("Today");
        }
        get label() {
            return this.y;
        }
        set label(a) {
            this.y = a;
            this.e("Label");
        }
        get labelTextColor() {
            return this.ac;
        }
        set labelTextColor(a) {
            this.ac = a;
            this.e("LabelTextColor");
        }
        get labelFontFamily() {
            return this.z;
        }
        set labelFontFamily(a) {
            this.z = a;
            this.e("LabelFontFamily");
        }
        get labelFontSize() {
            return this.r;
        }
        set labelFontSize(a) {
            this.r = a;
            this.e("LabelFontSize");
        }
        get labelFontStyle() {
            return this.aa;
        }
        set labelFontStyle(a) {
            this.aa = a;
            this.e("LabelFontStyle");
        }
        get labelFontWeight() {
            return this.ab;
        }
        set labelFontWeight(a) {
            this.ab = a;
            this.e("LabelFontWeight");
        }
        get placeholder() {
            return this.ad;
        }
        set placeholder(a) {
            this.ad = a;
            this.e("Placeholder");
        }
        get minDate() {
            return this.k;
        }
        set minDate(a) {
            this.k = a;
            this.e("MinDate");
        }
        get maxDate() {
            return this.j;
        }
        set maxDate(a) {
            this.j = a;
            this.e("MaxDate");
        }
        get density() {
            return this.w;
        }
        set density(a) {
            this.w = a;
            this.e("Density");
        }
        get baseTheme() {
            return this.v;
        }
        set baseTheme(a) {
            this.v = a;
            this.e("BaseTheme");
        }
        get textFontFamily() {
            return this.af;
        }
        set textFontFamily(a) {
            this.af = a;
            this.e("TextFontFamily");
        }
        get textFontSize() {
            return this.s;
        }
        set textFontSize(a) {
            this.s = a;
            this.e("TextFontSize");
        }
        get textFontStyle() {
            return this.ag;
        }
        set textFontStyle(a) {
            this.ag = a;
            this.e("TextFontStyle");
        }
        get textFontWeight() {
            return this.ah;
        }
        set textFontWeight(a) {
            this.ah = a;
            this.e("TextFontWeight");
        }
        get textColor() {
            return this.ae;
        }
        set textColor(a) {
            this.ae = a;
            this.e("TextColor");
        }
        get iconColor() {
            return this.x;
        }
        set iconColor(a) {
            this.x = a;
            this.e("IconColor");
        }
        get showClearButton() {
            return this.f;
        }
        set showClearButton(a) {
            this.f = a;
            this.e("ShowClearButton");
        }
        get showTodayButton() {
            return this.g;
        }
        set showTodayButton(a) {
            this.g = a;
            this.e("ShowTodayButton");
        }
        get valueChangedRef() {
            return this.ai;
        }
        set valueChangedRef(a) {
            this.ai = a;
            this.e("ValueChangedRef");
        }
    }
    DatePickerDescription.$t = /*@__PURE__*/ markType(DatePickerDescription, 'DatePickerDescription', Description.$);
    return DatePickerDescription;
})();
