// igniteui-webcomponents-core/esm2015/lib/LocalDataSource_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseDataSource } from "./BaseDataSource";
import { Base, typeCast, Array_$type, Number_$type, fromEnum, markType } from "./type";
import { RearrangedListView } from "./RearrangedListView";
import { DataSourcePropertiesComparer } from "./DataSourcePropertiesComparer";
import { List$1 } from "./List$1";
import { DataSourceExecutionContext } from "./DataSourceExecutionContext";
import { DataSourceLocalDataProvider } from "./DataSourceLocalDataProvider";
import { DataSourceSpecialRow } from "./DataSourceSpecialRow";
import { ISupportsDataChangeNotifications_$type } from "./ISupportsDataChangeNotifications";
import { TransactionType, TransactionType_$type } from "./TransactionType";
import { ISupportsExpansionChangeNotifications_$type } from "./ISupportsExpansionChangeNotifications";
/**
 * @hidden
 */
export let LocalDataSource = /*@__PURE__*/ (() => {
    class LocalDataSource extends BaseDataSource {
        constructor() {
            super();
            this.ct = null;
            this.cs = null;
            this.dt = null;
            this.cy = null;
            this.c2 = null;
            this.c4 = null;
            this.c0 = null;
            this.de = new List$1(Array_$type, 0);
            this.c5 = false;
            this.cr = null;
            this.dm = -1;
            this.df = -1;
            this.dp = -1;
            this.dh = -1;
            this.dq = -1;
            this.di = -1;
            this.dn = -1;
            this.dg = -1;
            this.ds = -1;
            this.dr = -1;
            this.dk = -1;
            this.dj = -1;
            this.dl = -1;
            this.executionContext = new DataSourceExecutionContext();
            this.t.l = false;
        }
        get_actualBaseDataProvider() {
            return this.actualDataProvider;
        }
        get actualBaseDataProvider() {
            return this.get_actualBaseDataProvider();
        }
        get dataProvider() {
            return this.ct;
        }
        set dataProvider(a) {
            let b = this.ct;
            this.ct = a;
            if (b != this.ct) {
                this.onPropertyUpdated("DataProvider", b, this.ct);
            }
        }
        get actualDataProvider() {
            if (this.cs == null) {
                this.actualDataProvider = this.resolveDataProvider();
            }
            return this.cs;
        }
        set actualDataProvider(a) {
            let b = this.cs;
            this.cs = a;
            if (b != this.cs) {
                this.onPropertyUpdated("ActualDataProvider", b, this.cs);
            }
        }
        get dataSource() {
            return this.dt;
        }
        set dataSource(a) {
            let b = this.dt;
            this.dt = a;
            if (b != this.dt) {
                this.onPropertyUpdated("ItemsSource", b, this.dt);
            }
        }
        canInitialize() {
            let a = super.canInitialize();
            if (!a) {
                return false;
            }
            return this.dataSource != null;
        }
        get cz() {
            return this.cy;
        }
        set cz(a) {
            this.cy = a;
            this.actualDataProvider = this.resolveDataProvider();
        }
        resolveDataProvider() {
            let provider_ = new DataSourceLocalDataProvider();
            provider_.executionContext = new DataSourceExecutionContext();
            return provider_;
        }
        propertyUpdatedOverride(a, b, c) {
            super.propertyUpdatedOverride(a, b, c);
            switch (a) {
                case "DataProvider":
                    this.actualDataProvider = this.dataProvider;
                    break;
                case "ActualDataProvider":
                    if (this.actualDataProvider != null) {
                        this.actualDataProvider.dataSource = this.dataSource;
                        this.actualSchema = this.actualDataProvider.actualSchema;
                        this.actualCount = this.resolveFullCount(this.actualDataProvider.actualCount);
                    }
                    this.queueAutoRefresh();
                    break;
                case "ItemsSource":
                    if (this.actualDataProvider != null) {
                        this.actualDataProvider.dataSource = this.dataSource;
                        this.actualSchema = this.actualDataProvider.actualSchema;
                        this.actualCount = this.resolveFullCount(this.actualDataProvider.actualCount);
                    }
                    this.queueAutoRefresh();
                    break;
            }
        }
        getItemAtIndex(a) {
            if (this.c2 != null && this.c2.a6 != null) {
                return this.c2.a6.item(a);
            }
            else {
                return this.actualDataProvider.getItemAtIndex(a);
            }
        }
        getItemFromKey(a) {
            let b = this.actualPrimaryKey;
            if (b == null || b.length < 1 || this.actualBaseDataProvider == null) {
                return a[0];
            }
            let c = this.c1();
            let d;
            if (this.c2 != null) {
                let e = this.c2.a6.count;
                for (let f = 0; f < e; f++) {
                    d = this.c2.a6.item(f);
                    if (typeCast(DataSourceSpecialRow.$, d) !== null) {
                        let g = d;
                        if (g.rowType == 1 && !this.isSectionContentVisible && this.isSectionHeaderNormalRow) {
                            if (c.i(d, a) == 0) {
                                return d;
                            }
                        }
                    }
                    else if (c.i(d, a) == 0) {
                        return d;
                    }
                }
                return null;
            }
            if (this.actualDataProvider != null && this.actualDataProvider.isKeyIndexLookupSupported) {
                let h = this.actualDataProvider.indexOfKey(a);
                return this.actualDataProvider.getItemAtIndex(h);
            }
            if (this.actualDataProvider != null) {
                let i = this.actualDataProvider.actualCount;
                for (let j = 0; j < i; j++) {
                    d = this.actualDataProvider.getItemAtIndex(j);
                    if (c.i(d, a) == 0) {
                        return d;
                    }
                }
            }
            return null;
        }
        getItemPropertyAtIndex(a, b) {
            let c;
            if (this.c2 != null) {
                c = this.c2.a6.item(a);
            }
            else {
                c = this.actualDataProvider.getItemAtIndex(a);
            }
            let item_ = c;
            if ((item_ != null && item_.$$isSpecialRow !== undefined)) {
                return this.getSpecialRowItemValue(c, b);
            }
            if (this.ao()) {
                let d = this.al(c, b);
                if (d) {
                    let e = this.ay(c, b);
                    return e;
                }
            }
            return this.actualDataProvider.getItemValue(c, b);
        }
        cx() {
            return this.actualDataProvider;
        }
        c7() {
            return this.de != null && this.de.count > 0;
        }
        refreshInternalOverride() {
            super.refreshInternalOverride();
            if (this.actualDataProvider != null) {
                if (this.c8() || this.c9() || this.da() || this.db() || this.c7()) {
                    if (this.c2 == null) {
                        if (this.c4 == null) {
                            this.c4 = new LocalDataSourceViewUpdateNotifier(this);
                        }
                        this.c2 = new RearrangedListView(this.actualDataProvider);
                        this.c2.ab = this.c4;
                        this.c2.ad = new LocalDataSourceViewExpansionNotifier(this);
                    }
                    this.ensureComparables(this.actualDataProvider.actualSchema);
                    this.c2.av = this.shouldEmitSectionHeaders && this.an;
                    this.c2.au = this.shouldEmitSectionFooters && this.an;
                    this.c2.aw = this.shouldEmitShiftedRows;
                    this.c2.cb(this.de);
                    this.c2.a2 = this.dd();
                    this.c2.a1 = this.dc();
                    this.c2.p = this.c1();
                    this.c2.ae = this.groupDescriptions;
                    this.c2.af = this.sortDescriptions;
                    this.c2.ag = this.summaryDescriptions;
                    this.c2.l = this.summaryScope;
                    this.c2.am = this.includeSummaryRowsInSection;
                    this.c2.as = this.isSectionSummaryRowsAtBottom;
                    this.c2.j = this.sectionHeaderDisplayMode;
                    this.c2.m = this;
                    this.c2.ap = this.groupDescriptions != null && this.groupDescriptions.k.count > 0;
                    this.c2.t = this.c3();
                    this.c2.b6();
                    this.actualSchema = this.actualDataProvider.actualSchema;
                    this.actualCount = this.c2.a6.count;
                }
                else {
                    let a = this.cx();
                    this.actualSchema = this.actualDataProvider.actualSchema;
                    if (a != null) {
                        this.actualCount = a.actualCount;
                    }
                    else {
                        this.actualCount = 0;
                    }
                    if (this.c2 != null) {
                        this.c2.ab = null;
                        this.c2 = null;
                        this.onBroadcastClearItems();
                    }
                }
            }
            else {
                this.actualCount = 0;
                this.actualSchema = null;
            }
        }
        get_isItemIndexLookupSupported() {
            return true;
        }
        get_isKeyIndexLookupSupported() {
            return true;
        }
        indexOfItem(a) {
            if (this.c2 != null) {
                return this.c2.a6.indexOf(a);
            }
            if (this.actualDataProvider != null && this.actualDataProvider.isItemIndexLookupSupported) {
                return this.actualDataProvider.indexOfItem(a);
            }
            if (this.actualDataProvider != null) {
                let b = this.actualDataProvider.actualCount;
                let c;
                for (let d = 0; d < b; d++) {
                    c = this.actualDataProvider.getItemAtIndex(d);
                    if (this.c6(c, a)) {
                        return d;
                    }
                }
            }
            return -1;
        }
        c6(a, b) {
            if (a == null && b == null) {
                return true;
            }
            if (a == null || b == null) {
                return false;
            }
            return Base.equalsStatic(a, b);
        }
        indexOfKey(a) {
            let b = this.actualPrimaryKey;
            if (b == null || b.length < 1 || this.actualBaseDataProvider == null) {
                return this.indexOfItem(a[0]);
            }
            let c = this.c1();
            let d;
            if (this.c2 != null) {
                let e = this.c2.a6.count;
                for (let f = 0; f < e; f++) {
                    d = this.c2.a6.item(f);
                    if (typeCast(DataSourceSpecialRow.$, d) !== null) {
                        let g = d;
                        if (g.rowType == 1 && !this.isSectionContentVisible && this.isSectionHeaderNormalRow) {
                            if (c.i(d, a) == 0) {
                                return f;
                            }
                        }
                    }
                    else if (c.i(d, a) == 0) {
                        return f;
                    }
                }
                return -1;
            }
            if (this.actualDataProvider != null && this.actualDataProvider.isKeyIndexLookupSupported) {
                return this.actualDataProvider.indexOfKey(a);
            }
            if (this.actualDataProvider != null) {
                let h = this.actualDataProvider.actualCount;
                for (let i = 0; i < h; i++) {
                    d = this.actualDataProvider.getItemAtIndex(i);
                    if (c.i(d, a) == 0) {
                        return i;
                    }
                }
            }
            return -1;
        }
        onActualPrimaryKeyChangedOverrride(a, b) {
            super.onActualPrimaryKeyChangedOverrride(a, b);
            this.c0 = null;
        }
        c1() {
            if (this.c0 != null) {
                return this.c0;
            }
            if (this.actualPrimaryKey == null) {
                this.c0 = new DataSourcePropertiesComparer(null, null, null, this);
                return this.c0;
            }
            let a = new Array(this.actualPrimaryKey.length);
            for (let b = 0; b < a.length; b++) {
                a[b] = 0;
            }
            this.c0 = new DataSourcePropertiesComparer(this.actualPrimaryKey, a, this.d, this);
            return this.c0;
        }
        c3() {
            if (this.filterExpressions.j.count == 0) {
                return null;
            }
            let a = this.actualDataProvider;
            let b = this.actualDataProvider.actualSchema;
            let c = this.filterExpressions.j;
            let d = (e) => {
                for (let f = 0; f < c.count; f++) {
                    if (!c.item(f).match(e, a, b)) {
                        return false;
                    }
                }
                return true;
            };
            return d;
        }
        dd() {
            let a = 0;
            if (this.sortDescriptions.k.count == 0 && (this.groupDescriptions == null || this.groupDescriptions.k.count == 0)) {
                return null;
            }
            a = this.sortDescriptions.k.count;
            if (this.groupDescriptions != null) {
                a += this.groupDescriptions.k.count;
            }
            let b = new Array(a);
            let c = new Array(a);
            let d = 0;
            if (this.groupDescriptions != null) {
                for (let e = 0; e < this.groupDescriptions.k.count; e++) {
                    b[e] = this.groupDescriptions.k.item(e).propertyName;
                    c[e] = this.groupDescriptions.k.item(e).direction;
                    d++;
                }
            }
            for (let f = 0; f < this.sortDescriptions.k.count; f++) {
                b[f + d] = this.sortDescriptions.k.item(f).propertyName;
                c[f + d] = this.sortDescriptions.k.item(f).direction;
            }
            return new DataSourcePropertiesComparer(b, c, this.d, this);
        }
        dc() {
            if (this.groupDescriptions == null || this.groupDescriptions.k.count < 1) {
                return null;
            }
            let a = 0;
            a = this.groupDescriptions.k.count;
            let b = new Array(a);
            let c = new Array(a);
            if (this.groupDescriptions != null) {
                for (let d = 0; d < this.groupDescriptions.k.count; d++) {
                    b[d] = this.groupDescriptions.k.item(d).propertyName;
                    c[d] = this.groupDescriptions.k.item(d).direction;
                }
            }
            return new DataSourcePropertiesComparer(b, c, this.d, this);
        }
        c9() {
            if (this.groupDescriptions == null || this.groupDescriptions.k.count < 1) {
                return false;
            }
            if (this.actualDataProvider == null) {
                return false;
            }
            if (this.actualDataProvider.isGroupingSupported) {
                return false;
            }
            return true;
        }
        da() {
            if (this.sortDescriptions == null || this.sortDescriptions.k.count < 1) {
                return false;
            }
            if (this.actualDataProvider == null) {
                return false;
            }
            if (this.actualDataProvider.isSortingSupported) {
                return false;
            }
            return true;
        }
        c8() {
            return this.filterExpressions != null && this.filterExpressions.j.count > 0;
        }
        db() {
            if (this.summaryDescriptions == null || this.summaryDescriptions.k.count < 1) {
                return false;
            }
            return true;
        }
        onClearItemsOverride() {
            if (this.c2 != null) {
                this.c2.notifyClearItems();
                this.queueAutoRefresh();
            }
            else {
                super.onClearItemsOverride();
            }
        }
        onSetItemOverride(a, b, c) {
            if (this.c2 != null) {
                this.c2.notifySetItem(a, b, c);
            }
            else {
                super.onSetItemOverride(a, b, c);
            }
        }
        onInsertItemOverride(a, b) {
            if (this.c2 != null) {
                this.c2.notifyInsertItem(a, b);
            }
            else {
                super.onInsertItemOverride(a, b);
            }
        }
        onRemoveItemOverride(a, b) {
            if (this.c2 != null) {
                this.c2.notifyRemoveItem(a, b);
            }
            else {
                super.onRemoveItemOverride(a, b);
            }
        }
        dy() {
            this.actualCount = this.c2.a6.count;
            this.onBroadcastClearItems();
        }
        d1(a, b, c) {
            this.actualCount = this.c2.a6.count;
            this.onBroadcastSetItem(a, b, c);
        }
        dz(a, b) {
            this.actualCount = this.c2.a6.count;
            this.onBroadcastInsertItem(a, b);
        }
        d0(a, b) {
            this.actualCount = this.c2.a6.count;
            this.onBroadcastRemoveItem(a, b);
        }
        pinRow(a) {
            super.pinRow(a);
            this.dm = -1;
            this.df = -1;
            if (this.c2 != null) {
                this.de.add(a);
                this.c2.b5(a);
            }
            else {
                this.de.add(a);
                this.queueAutoRefresh();
            }
        }
        unpinRow(a) {
            super.unpinRow(a);
            let b = this.c1();
            if (this.c2 != null) {
                for (let c = this.de.count - 1; c >= 0; c--) {
                    let d = this.de._inner[c];
                    if (d.length != a.length) {
                        continue;
                    }
                    if (b.h(d, a) == 0) {
                        this.de.removeAt(c);
                        this.c2.cc(d);
                    }
                }
            }
            else {
                for (let e = this.de.count - 1; e >= 0; e--) {
                    let f = this.de._inner[e];
                    if (f.length != a.length) {
                        continue;
                    }
                    if (b.h(f, a) == 0) {
                        this.de.removeAt(e);
                    }
                }
                this.queueAutoRefresh();
            }
            this.dm = -1;
            this.df = -1;
        }
        clearPinnedRows() {
            for (let a = this.de.count - 1; a >= 0; a--) {
                this.unpinRow(this.de._inner[a]);
            }
        }
        isRowPinned(a) {
            if (this.c2 == null) {
                return false;
            }
            let b = this.getItemAtIndex(a);
            if (b == null) {
                return false;
            }
            let item_ = b;
            if ((item_ != null && item_.$$isSpecialRow !== undefined)) {
                if (item_.rowType == 3) {
                    return true;
                }
            }
            let c = this.c1();
            let d;
            if (this.actualPrimaryKey == null) {
                d = new Array(1);
                d[0] = b;
            }
            else {
                let e = this.actualPrimaryKey;
                let f = e.length;
                d = new Array(f);
                for (let g = 0; g < f; g++) {
                    d[g] = this.getItemProperty(b, e[g]);
                }
            }
            for (let h = 0; h < this.de.count; h++) {
                if (c.h(this.de._inner[h], d) == 0) {
                    return true;
                }
            }
            return false;
        }
        get_isSortingSupported() {
            if (this.externalDataSource != null && !this.c5) {
                this.c5 = true;
                let a = this.externalDataSource.isSortingSupportedOverride;
                this.c5 = false;
                return a;
            }
            return true;
        }
        get_isFilteringSupported() {
            if (this.externalDataSource != null && !this.c5) {
                this.c5 = true;
                let a = this.externalDataSource.isFilteringSupportedOverride;
                this.c5 = false;
                return a;
            }
            return true;
        }
        get_isGroupingSupported() {
            if (this.externalDataSource != null && !this.c5) {
                this.c5 = true;
                let a = this.externalDataSource.isGroupingSupportedOverride;
                this.c5 = false;
                return a;
            }
            return true;
        }
        getStickyRowPriority(a) {
            let b = this.getRowType(a);
            switch (b) {
                case 1: return 100 - this.getRowLevel(a);
                case 2: return 99 - this.getRowLevel(a);
                case 3:
                case 5: return 2;
                case 4: return 3;
            }
            return 1;
        }
        getStickyRowsInRange(a, b) {
            if (this.c2 == null || ((this.groupDescriptions == null || this.groupDescriptions.k.count == 0) && (this.summaryDescriptions == null || this.summaryDescriptions.k.count == 0) && !this.c7())) {
                return null;
            }
            let c = null;
            let d = null;
            if (this.c2.a5.count == 1) {
                c = this.c2.a5._inner[0];
                d = this.c2.a5._inner[0];
            }
            else {
                c = this.c2.w(a);
                if (c != null) {
                    while (c.z != -1) {
                        c = this.c2.a5._inner[c.z];
                    }
                }
                d = this.c2.w(b);
                if (b >= this.c2.a6.count - this.c2.bk) {
                    d = this.c2.v();
                }
            }
            if (c == null || d == null) {
                return null;
            }
            if (c.ae == this.dm && c.aa == this.dp && c.af == this.dq && c.ad == this.dn && c.ah == this.ds && c.ag == this.dr && d.ae == this.df && d.aa == this.dh && d.af == this.di && d.ad == this.dg && d.ah == this.dk && d.ag == this.dj && this.c2.bk == this.dl) {
                return this.cr;
            }
            this.dm = c.ae;
            this.df = d.ae;
            this.dp = c.aa;
            this.dh = d.aa;
            this.dq = c.af;
            this.di = d.af;
            this.dn = c.ad;
            this.dg = d.ad;
            this.ds = c.ah;
            this.dr = c.ag;
            this.dk = d.ah;
            this.dj = d.ag;
            this.dl = this.c2.bk;
            if (this.c2.a5.count == 1) {
                this.df = d.ad;
            }
            let e = 0;
            if (this.shouldEmitSectionHeaders && this.an) {
                e = 1;
            }
            let f = this.c2.ba();
            let g = new List$1(Number_$type, 0);
            for (let h = c.ae; h <= d.ae; h++) {
                let i = this.c2.a5._inner[h];
                let j = this.c2.a5._inner[h].c.length;
                let k = this.c2.a5._inner[h].v();
                if (this.shouldEmitSectionHeaders && this.an) {
                    for (let l = 0; l < j; l++) {
                        if (this.c2.a5._inner[h].a[l] != null) {
                            let m = i.af + i.b[l];
                            g.add(m);
                            if (this.summaryScope == 0 || this.summaryScope == 2) {
                                if (i.c[l] && this.includeSummaryRowsInSection && !this.isSectionSummaryRowsAtBottom) {
                                    for (let n = 0; n < f; n++) {
                                        g.add(m + 1 + n);
                                    }
                                }
                            }
                        }
                        if (!this.c2.a5._inner[h].c[l]) {
                            break;
                        }
                    }
                }
                if (this.c2.a5._inner[h].p) {
                    if (this.shouldEmitShiftedRows) {
                        for (let o = 0; o < this.c2.a5._inner[h].aa; o++) {
                            g.add(this.c2.a5._inner[h].af + this.c2.a5._inner[h].ah + o + k);
                        }
                    }
                }
                if (this.summaryScope == 0 || this.summaryScope == 2) {
                    if (this.includeSummaryRowsInSection && this.isSectionSummaryRowsAtBottom) {
                        for (let p = this.c2.a5._inner[h].ag - 1; p >= 0; p--) {
                            if (this.shouldEmitSectionFooters) {
                                g.add(this.c2.a5._inner[h].ad - k - p);
                            }
                            else {
                                g.add(this.c2.a5._inner[h].ad - p);
                            }
                        }
                    }
                }
                if (this.shouldEmitSectionFooters && this.an) {
                    for (let q = j - 1; q >= 0; q--) {
                        if (!this.c2.a5._inner[h].c[q]) {
                            break;
                        }
                        g.add(this.c2.a5._inner[h].ad - q);
                    }
                }
            }
            for (let r = this.c2.a6.count - this.c2.bk; r < this.c2.a6.count; r++) {
                g.add(r);
            }
            this.cr = g.toArray();
            return this.cr;
        }
        isExclusivelySticky(a) {
            let b = this.getRowType(a);
            return b == 1 || b == 2 || b == 4 || (this.isSectionSummaryRowsAtBottom && b == 5);
        }
        getRowType(a) {
            if (a < 0) {
                return 0;
            }
            let b = this.getItemAtIndex(a);
            let item_ = b;
            if ((item_ != null && item_.$$isSpecialRow !== undefined)) {
                let c = b.rowType;
                if (this.isSectionHeaderNormalRow && c == 1) {
                    return 0;
                }
                return c;
            }
            return 0;
        }
        setIsRowExpandedAtIndex(a, b) {
            if (this.isSectionCollapsable && this.c2 != null && this.c2.a6.count > 0 && this.c2.a6.count > a) {
                this.c2.ca(a, b);
            }
        }
        getIsRowExpandedAtIndex(a) {
            if (this.c2 != null && this.c2.a6.count > 0 && this.c2.a6.count > a) {
                return this.c2.ak(a);
            }
            return true;
        }
        getRowLevel(a) {
            if (a < 0) {
                return 0;
            }
            let b = this.getItemAtIndex(a);
            let item_ = b;
            if ((item_ != null && item_.$$isSpecialRow !== undefined)) {
                return b.level;
            }
            if (this.sectionHeaderDisplayMode == 1) {
                return this.groupDescriptions.k.count;
            }
            else if (this.groupDescriptions.k.count > 0) {
                return 1;
            }
            return 0;
        }
        getRootSummaryRowCount() {
            if (this.c2 != null) {
                return this.c2.bk;
            }
            return 0;
        }
        getRootSummaryResults() {
            if (this.c2 != null) {
                return this.c2.i;
            }
            return null;
        }
        getSectionSummaryResults(a) {
            if (this.c2 != null && this.c2.a5.count > 0 && a >= 0 && a < this.c2.a5.count) {
                return this.c2.a5._inner[a].i;
            }
            return null;
        }
        clone() {
            let a = new LocalDataSource();
            a.dataSource = this.dataSource;
            super.cloneProperties(a);
            return a;
        }
        ao() {
            return !this.isReadOnly && this.actualBaseDataProvider != null && this.t != null;
        }
        ck(a, b) {
            if (a == 5) {
                let c = new List$1(TransactionType_$type, 0);
                let d = new List$1(Base.$, 0);
                let e = b;
                let f = e.a;
                for (let g = 0; g < f.length; g++) {
                    if (f[g].transactionType == 1) {
                        let h = f[g].id;
                        let i = this.getItemFromKey(h);
                        if (i != null) {
                            let j = this.indexOfItem(i);
                            let k = null;
                            if (this.c2 != null && j > -1) {
                                k = this.c2.w(j);
                                this.c2.ci(j, i, false, k);
                            }
                            let l = f[g].value;
                            for (let m of fromEnum(l)) {
                                this.actualDataProvider.setItemValue(i, m.key, m.value);
                            }
                            if (k != null && this.c2 != null && j > -1) {
                                this.c2.ci(j, i, true, k);
                            }
                            if (j > -1) {
                                this.onBroadcastSetItem(j, i, i);
                            }
                            d.add1(i);
                            c.add(1);
                            if (this.de.count > 0) {
                                let n = this.c1();
                                for (let o = 0; o < this.de.count; o++) {
                                    if (n.h(this.de._inner[o], h) == 0) {
                                        let p = this.getStickyRows();
                                        for (let q = 0; q < p.length; q++) {
                                            let r = p[q];
                                            let s = typeCast(DataSourceSpecialRow.$, this.c2.a6.item(r));
                                            if (s != null && s.rowType == 3 && s.targetRow == i) {
                                                this.onBroadcastSetItem(r, s, s);
                                                break;
                                            }
                                        }
                                        break;
                                    }
                                }
                            }
                        }
                    }
                    else if (f[g].transactionType == TransactionType.Delete) {
                        let t = f[g].id;
                        let u = this.indexOfKey(t);
                        if (u > -1) {
                            let v = this.getItemAtIndex(u);
                            this.actualDataProvider.removeItem(v);
                            this.notifyRemoveItem(u, v);
                            d.add1(v);
                            c.add(TransactionType.Delete);
                        }
                    }
                    else if (f[g].transactionType == 0) {
                        let w = f[g].id[0];
                        if (f[g].value != null) {
                            let x = f[g].value;
                            for (let y of fromEnum(x)) {
                                this.actualDataProvider.setItemValue(w, y.key, y.value);
                            }
                        }
                        this.actualDataProvider.addItem(w);
                        d.add1(w);
                        c.add(0);
                    }
                }
                if (f.length > 0) {
                    this.bw(e.id, this.getAggregatedChanges(e.id));
                }
            }
        }
    }
    LocalDataSource.$t = /*@__PURE__*/ markType(LocalDataSource, 'LocalDataSource', BaseDataSource.$);
    return LocalDataSource;
})();
/**
 * @hidden
 */
export let LocalDataSourceViewUpdateNotifier = /*@__PURE__*/ (() => {
    class LocalDataSourceViewUpdateNotifier extends Base {
        constructor(a) {
            super();
            this.a = null;
            this.a = a;
        }
        notifyClearItems() {
            this.a.dy();
        }
        notifyInsertItem(a, b) {
            this.a.dz(a, b);
        }
        notifyRemoveItem(a, b) {
            this.a.d0(a, b);
        }
        notifySetItem(a, b, c) {
            this.a.d1(a, b, c);
        }
    }
    LocalDataSourceViewUpdateNotifier.$t = /*@__PURE__*/ markType(LocalDataSourceViewUpdateNotifier, 'LocalDataSourceViewUpdateNotifier', Base.$, [ISupportsDataChangeNotifications_$type]);
    return LocalDataSourceViewUpdateNotifier;
})();
/**
 * @hidden
 */
export let LocalDataSourceViewExpansionNotifier = /*@__PURE__*/ (() => {
    class LocalDataSourceViewExpansionNotifier extends Base {
        constructor(a) {
            super();
            this.a = null;
            this.a = a;
        }
        notifyRowExpansionChanged(a, b, c) {
            this.a.onRowExpansionChanged(a, b, c);
        }
    }
    LocalDataSourceViewExpansionNotifier.$t = /*@__PURE__*/ markType(LocalDataSourceViewExpansionNotifier, 'LocalDataSourceViewExpansionNotifier', Base.$, [ISupportsExpansionChangeNotifications_$type]);
    return LocalDataSourceViewExpansionNotifier;
})();
