// igniteui-webcomponents-core/esm5/lib/DatePickerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DatePickerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(DatePickerDescription, _super);
    function DatePickerDescription() {
        var _this = _super.call(this) || this;
        _this.m = new Date();
        _this.l = new Date();
        _this.y = null;
        _this.ac = null;
        _this.z = null;
        _this.r = 0;
        _this.aa = null;
        _this.ab = null;
        _this.ad = null;
        _this.k = new Date();
        _this.j = new Date();
        _this.w = null;
        _this.v = null;
        _this.af = null;
        _this.s = 0;
        _this.ag = null;
        _this.ah = null;
        _this.ae = null;
        _this.x = null;
        _this.f = false;
        _this.g = false;
        _this.ai = null;
        return _this;
    }
    DatePickerDescription.prototype.get_type = function () {
        return "DatePicker";
    };
    Object.defineProperty(DatePickerDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "value", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.e("Value");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "today", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.e("Today");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "label", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.e("Label");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "labelTextColor", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.e("LabelTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "labelFontFamily", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.e("LabelFontFamily");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "labelFontSize", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.e("LabelFontSize");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "labelFontStyle", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.e("LabelFontStyle");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "labelFontWeight", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.e("LabelFontWeight");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "placeholder", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.e("Placeholder");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "minDate", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.e("MinDate");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "maxDate", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("MaxDate");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "density", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.e("Density");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "baseTheme", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.e("BaseTheme");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "textFontFamily", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.e("TextFontFamily");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "textFontSize", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.e("TextFontSize");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "textFontStyle", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.e("TextFontStyle");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "textFontWeight", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.e("TextFontWeight");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "textColor", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.e("TextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "iconColor", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.e("IconColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "showClearButton", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("ShowClearButton");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "showTodayButton", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
            this.e("ShowTodayButton");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DatePickerDescription.prototype, "valueChangedRef", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.e("ValueChangedRef");
        },
        enumerable: true,
        configurable: true
    });
    DatePickerDescription.$t = markType(DatePickerDescription, 'DatePickerDescription', Description.$);
    return DatePickerDescription;
}(Description));
export { DatePickerDescription };
